tests/benchmarks/sched_queues/src/main.c

```diff
@@ -0,0 +1,308 @@
+{
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	while (1) {
+	}
+}
+#endif
+
+/**
+ * The test entry routine is not expected to execute.
+ */
+static void test_entry(void *p1, void *p2, void *p3)
+{
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	printk("Thread %u unexpectedly executed\n",
+	       (unsigned int)(uintptr_t)p1);
+
+	while (1) {
+	}
+}
+
+static void start_threads(unsigned int num_threads)
+{
+	unsigned int i;
+	unsigned int bucket_size;
+
+#if CONFIG_MP_MAX_NUM_CPUS > 1
+
+	/* Start the busy threads to execute on the other processors */
+
+	for (i = 0; i < CONFIG_MP_MAX_NUM_CPUS - 1; i++) {
+		k_thread_create(&busy_thread[i], busy_stack[i], BUSY_STACK_SIZE,
+				busy_entry, NULL, NULL, NULL,
+				-1, 0, K_NO_WAIT);
+	}
+#endif
+
+	bucket_size = (num_threads / CONFIG_NUM_PREEMPT_PRIORITIES) + 1;
+
+	for (i = 0; i < CONFIG_BENCHMARK_NUM_THREADS; i++) {
+		k_thread_create(&test_thread[i], test_stack, TEST_STACK_SIZE,
+				test_entry, (void *)(uintptr_t)i, NULL, NULL,
+				i / bucket_size, 0, K_NO_WAIT);
+	}
+}
+
+static void cycles_reset(unsigned int num_threads)
+{
+	unsigned int i;
+
+	for (i = 0; i < num_threads; i++) {
+		add_cycles[i] = 0ULL;
+		remove_cycles[i] = 0ULL;
+	}
+}
+
+void test_decreasing_priority(unsigned int num_threads)
+{
+	unsigned int i;
+	timing_t start;
+	timing_t finish;
+
+	for (i = 0; i < num_threads; i++) {
+		start = timing_counter_get();
+		z_unready_thread(&test_thread[num_threads - i - 1]);
+		finish = timing_counter_get();
+		remove_cycles[num_threads - i - 1] += timing_cycles_get(&start, &finish);
+	}
+
+	for (i = 0; i < num_threads; i++) {
+		start = timing_counter_get();
+		z_ready_thread(&test_thread[i]);
+		finish = timing_counter_get();
+		add_cycles[i] += timing_cycles_get(&start, &finish);
+	}
+}
+
+void test_increasing_priority(unsigned int num_threads)
+{
+	unsigned int i;
+	timing_t start;
+	timing_t finish;
+
+	for (i = num_threads; i > 0; i--) {
+		start = timing_counter_get();
+		z_unready_thread(&test_thread[num_threads - i]);
+		finish = timing_counter_get();
+		remove_cycles[i - 1] += timing_cycles_get(&start, &finish);
+	}
+
+	for (i = num_threads; i > 0; i--) {
+		start = timing_counter_get();
+		z_ready_thread(&test_thread[i - 1]);
+		finish = timing_counter_get();
+		add_cycles[num_threads - i] += timing_cycles_get(&start, &finish);
+	}
+}
+
+void compute_and_report_stats(unsigned int num_threads,
+			      unsigned int num_iterations,
+			      uint64_t *cycles,
+			      const char *str)
+{
+	uint64_t minimum = cycles[0];
+	uint64_t maximum = cycles[0];
+	uint64_t total = cycles[0];
+	uint64_t average;
+
+	for (unsigned int i = 1; i < num_threads; i++) {
+		if (cycles[i] > maximum) {
+			maximum = cycles[i];
+		}
+
+		if (cycles[i] < minimum) {
+			minimum = cycles[i];
+		}
+
+		total += cycles[i];
+	}
+
+	minimum /= (uint64_t)num_iterations;
+	maximum /= (uint64_t)num_iterations;
+	average = total / (num_threads * num_iterations);
+
+	printk("%s\n", str);
```

Comment:
Please also compute the variance and stddev which are useful for seeing how wide the curve is around the mean (avg).

Reply:
Well, variance **or** standard deviation.   They're different conventions for measuring the same thing and can be trivially computed from each other.

Reply:
Can do.